src/js/components/conversation.jsx

```diff
@@ -54,34 +66,146 @@ export class ConversationComponent extends Component {
+
+        const node = findDOMNode(this);
+        if (hasMoreMessages && !isFetchingMoreMessages) {
+            // make sure the last message is one from the server, otherwise it doesn't need to scroll to previous first message
+            if (messages.length > 0 && messages[messages.length - 1]._id) {
+                this._lastTopMessageId = messages[0]._id;
+            }
+            
+            const top = getTop(this._topMessageNode, node);
+            this._lastTopMessageNodePosition = top - node.scrollTop;
+            dispatch(setFetchingMoreMessages(true));
+            setTimeout(() => {
+                fetchMoreMessages();
+            }, 400);
+        }
+    };
+
     scrollToBottom = () => {
-        const timeout = setTimeout(() => {
+        const {shouldScrollToBottom} = this.props;
+        if(!this._isScrolling && (shouldScrollToBottom || this._forceScrollToBottom)) {
+            this._isScrolling = true;
+            const timeout = setTimeout(() => {
+                const container = findDOMNode(this);
+                const logo = this.refs.logo;
+                const scrollTop = container.scrollHeight - container.clientHeight - logo.clientHeight - INTRO_BOTTOM_SPACER;
+                container.scrollTop = scrollTop;
+                this._isScrolling = false;
+                this._forceScrollToBottom = false;
+            });
+            this.scrollTimeouts.push(timeout);
+        }
+    };
+
+    scrollToPreviousFirstMessage = (node = undefined) => {
+        // This will scroll to specified node if we've reached the oldest messages.
+        // Otherwise, scroll to this._lastTopMessageNode
+        if (node) {
             const container = findDOMNode(this);
-            const logo = this.refs.logo;
-            const scrollTop = container.scrollHeight - container.clientHeight - logo.clientHeight - INTRO_BOTTOM_SPACER;
-            container.scrollTop = scrollTop;
-        });
-        this.scrollTimeouts.push(timeout);
+            container.scrollTop = getTop(node, container) - LOAD_MORE_LINK_HEIGHT;
+        } else {
+            if (this._lastTopMessageNodePosition && !this._isScrolling) {
+                const container = findDOMNode(this);
+                const node = this._lastTopMessageNode;
+                this._isScrolling = true;
+
+                // When fetching more messages, we want to make sure that after 
+                // render, the messages stay in the same places
+                container.scrollTop = getTop(node, container) - this._lastTopMessageNodePosition;
+
+                const timeout = setTimeout(() => {
+                    this._isScrolling = false;
+                });
+
+                this.scrollTimeouts.push(timeout);
+            }
+        }
+        this._lastTopMessageNode = undefined;
+        
     };
 
+    componentWillUpdate(nextProps) {
+        const {messages: currentMessages, isFetchingMoreMessages} = this.props;
```

Comment:
I put this in `componentWillUpdate` because I wanted to check for a newly received message. `componentWillUpdate` gives me access to the new props and the old props. I compare the length of the messages in the props to see if a new message was sent.